src/openrct2/scripting/ScSocketServer.hpp

```diff
@@ -0,0 +1,505 @@
+        }
+
+    public:
+        void Raise(
+            uint32_t id, const std::shared_ptr<Plugin>& plugin, const std::vector<DukValue>& args, bool isGameStateMutable)
+        {
+            auto& scriptEngine = GetContext()->GetScriptEngine();
+
+            // Use simple for i loop in case listeners is modified during the loop
+            auto listeners = GetListenerList(id);
+            for (size_t i = 0; i < listeners.size(); i++)
+            {
+                scriptEngine.ExecutePluginCall(plugin, listeners[i], args, isGameStateMutable);
+
+                // Safety, listeners might get reallocated
+                listeners = GetListenerList(id);
+            }
+        }
+
+        void AddListener(uint32_t id, const DukValue& listener)
+        {
+            auto& listeners = GetListenerList(id);
+            listeners.push_back(listener);
+        }
+
+        void RemoveListener(uint32_t id, const DukValue& value)
+        {
+            auto& listeners = GetListenerList(id);
+            auto it = std::find(listeners.begin(), listeners.end(), value);
+            if (it != listeners.end())
+            {
+                listeners.erase(it);
+            }
+        }
+
+        void RemoveAllListeners(uint32_t id)
+        {
+            auto& listeners = GetListenerList(id);
+            listeners.clear();
+        }
+    };
+
+    class ScSocketBase
+    {
+    private:
+        std::shared_ptr<Plugin> _plugin;
+
+    protected:
+        static bool IsLocalhostAddress(const std::string_view& s)
+        {
+            return s == "localhost" || s == "127.0.0.1" || s == "::";
+        }
+
+    public:
+        ScSocketBase(const std::shared_ptr<Plugin>& plugin)
+            : _plugin(plugin)
+        {
+        }
+
+        virtual ~ScSocketBase()
+        {
+            Dispose();
+        }
+
+        const std::shared_ptr<Plugin>& GetPlugin() const
+        {
+            return _plugin;
+        }
+
+        virtual void Update() = 0;
+
+        virtual void Dispose()
+        {
+        }
+
+        virtual bool IsDisposed() const = 0;
+    };
+
+    class ScSocket : public ScSocketBase
+    {
+    private:
+        static constexpr uint32_t EVENT_NONE = std::numeric_limits<uint32_t>::max();
+        static constexpr uint32_t EVENT_CLOSE = 0;
+        static constexpr uint32_t EVENT_DATA = 1;
+        static constexpr uint32_t EVENT_CONNECT_ONCE = 2;
+        static constexpr uint32_t EVENT_ERROR = 3;
+
+        EventList _eventList;
+        std::unique_ptr<ITcpSocket> _socket;
+        bool _disposed{};
+        bool _connecting{};
+
+    public:
+        ScSocket(const std::shared_ptr<Plugin>& plugin)
+            : ScSocketBase(plugin)
+        {
+        }
+
+        ScSocket(const std::shared_ptr<Plugin>& plugin, std::unique_ptr<ITcpSocket>&& socket)
+            : ScSocketBase(plugin)
+            , _socket(std::move(socket))
+        {
+        }
+
+    private:
+        ScSocket* destroy(const DukValue& error)
```

Comment:
I would just call this close instead of destroy its a bit misleading between dispose/destroy.

Reply:
This is mirroring the node.js API, so unfortunately not something I'd like to change.